torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+    size_with_reads: int = 0  # this is used only in topological_sort_dfs
+
+
+@dataclasses.dataclass
+class FreeableInputBuffer:
+    dep: Dep
+    mpi_buffer: MemoryPlanningInfoForBuffer = dataclasses.field(
+        default_factory=MemoryPlanningInfoForBuffer
+    )
+
+    def get_name(self) -> str:
+        return self.dep.name
+
+    def __hash__(self) -> int:
+        return hash(self.dep.name)
+
+
+def dep_size_hint(dep: Dep) -> int:
+    res = 0
+    try:
+        if not dep.has_unbacked_symbols():
+            res = dep.numbytes_hint()
+    except KeyError:
+        # In at least one test (test/inductor/test_torchbind.py) we
+        # create a StarDep that doesn't exist in the graph and calling
+        # `has_unbacked_symbols()` throws an error.
+        pass
+    return res
+
+
+def get_freeable_input_buf(
+    nodes: List[BaseSchedulerNode],
+    graph_inputs: Set[str],
+) -> Dict[str, FreeableInputBuffer]:
+    """
+    Create and keep track of all input buffers that can be freed during the program
+    """
+    name_to_input_buf: Dict[str, FreeableInputBuffer] = {}
+    for node in nodes:
+        for dep in node.read_writes.reads:
+            if (
+                dep.name in graph_inputs
+                and not dep.name.startswith("primals_")
+                and dep.name not in name_to_input_buf
+            ):
+                name_to_input_buf[dep.name] = FreeableInputBuffer(dep)
+                name_to_input_buf[dep.name].mpi_buffer.size_free = dep_size_hint(dep)
+
+    return name_to_input_buf
+
+
+def compute_size_for_scheduler_buffer(name_to_buf: Dict[str, SchedulerBuffer]) -> None:
+    """
+    Compute the size of each scheduler buffer, including (1) memory allocated when
+    it is created and (2) memory deallocated when it is freed.
+
+    We specially handle the case of MultiOutputLayout.
+    Consider the following case:
+        buf0 = some_ops_with_multi_outputs(...)
+        buf1 = buf0[0] # assume 10 bytes
+        buf2 = buf0[1] # assume 20 bytes
+    In such cases,
+        buf0: at creation, 30 bytes allocated, when deleted, 0 bytes freed
+        buf1: at creation, 0 bytes allocated, when deleted, 10 bytes freed
+        buf2: at creation, 0 bytes allocated, when deleted, 20 bytes freed
+    """
+    from .scheduler import BaseSchedulerNode, OutputNode
+
+    # compute the size of SchedulerBuffer without MultiOutputLayout layout
+    for sched_buf in name_to_buf.values():
+        if not isinstance(sched_buf.node.layout, MultiOutputLayout):
+            sched_buf.mpi_buffer = MemoryPlanningInfoForBuffer()
+            sched_buf.mpi_buffer.size_alloc = V.graph.sizevars.size_hint(
+                sched_buf.node.get_numel(), fallback=0
+            ) * get_dtype_size(sched_buf.node.get_dtype())
+            sched_buf.mpi_buffer.size_free = sched_buf.mpi_buffer.size_alloc
+
+    # compute the size of SchedulerBuffer with MultiOutputLayout layout
+    for sched_buf in name_to_buf.values():
```

Comment:
I need to process the size for SchedulerBuffer without MultiOutputLayout layout first before doing so for those with MultiOutputLayout layout. So that's why I kept two separate loops.